pkg/apis/pipeline/v1/matrix_types.go

```diff
@@ -64,13 +62,85 @@ type Combinations []Combination
 
 // FanOut returns an list of params that represent combinations
 func (m *Matrix) FanOut() []Params {
-	var combinations Combinations
+	var combinations, includeCombinations Combinations
+	includeCombinations = m.getIncludeCombinations()
+	if m.HasInclude() && !m.HasParams() {
+		// If there are only Matrix Include Parameters return explicit combinations
+		return includeCombinations.toParams()
+	}
+	// Generate combinations from Matrix Parameters
 	for _, parameter := range m.Params {
 		combinations = combinations.fanOutMatrixParams(parameter)
 	}
+	combinations = combinations.overwriteCombinations(includeCombinations)
+	combinations = combinations.addNewCombinations(includeCombinations)
 	return combinations.toParams()
 }
 
+// overwriteCombinations replaces any missing include params in the initial
+// matrix params combinations by overwriting the initial combinations with the
+// include combinations
+func (cs Combinations) overwriteCombinations(ics Combinations) Combinations {
+	for _, paramCombination := range cs {
+		for _, includeCombination := range ics {
+			if paramCombination.contains(includeCombination) {
+				includeCombination.overwrite(paramCombination)
+			}
+		}
+	}
+	return cs
+}
+
+// addNewCombinations creates a new combination for any include parameter
+// values that are missing entirely from the initial combinations and
+// returns all combinations
+func (cs Combinations) addNewCombinations(ics Combinations) Combinations {
+	for _, includeCombination := range ics {
+		if cs.shouldAddNewCombination(includeCombination) {
+			cs = append(cs, includeCombination)
+		}
+	}
+	return cs
+}
+
+// contains returns true if the include parameter name and value exists in combinations
+func (c Combination) contains(includeCombination Combination) bool {
+	for name, val := range includeCombination {
+		if _, exist := c[name]; exist {
+			if c[name] != val {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// overwrite the parameter name and value exists in combination with the include combination
+func (c Combination) overwrite(oldCombination Combination) Combination {
```

Comment:
NIT: this is creating unnecessary call chain, making it confusing to understand, not a blocker though

Reply:
@pritidesai A PR has been opened separately here: https://github.com/tektoncd/pipeline/pull/6463